### PR TITLE
View and restore previously saved qsets in creator

### PR DESCRIPTION
**Pull request by clpetersonucf**
To be paired with https://github.com/ucfopen/Materia-Server-Client-Assets/pull/97

### CODE REVIEW

**Comment by clpetersonucf** [comment]
Ready for initial review.

File: fuel/app/classes/materia/api/v1.php [hunk 1]
```diff
@@ -775,6 +775,17 @@ static public function question_set_get($inst_id, $play_id = null, $timestamp =
 		return $inst->qset;
 	}
 
+	static public function question_history_get($inst_id)
```
Comment (iturgeon) [inline fuel/app/classes/materia/api/v1.php:778]: Would prefer any new api endpoints to be added via fuel rest controllers, not sure if that's reasonable here or not.

Comment (iturgeon) [reply fuel/app/classes/materia/api/v1.php:778]: Add type hinting to function params and return

File: fuel/app/classes/materia/widget/instance.php [hunk 2]
```diff
@@ -180,6 +180,31 @@ public function get_qset($inst_id, $timestamp=false)
 		}
 	}
 
+	public function get_qset_history($inst_id)
```
Comment (iturgeon) [inline fuel/app/classes/materia/widget/instance.php:183]: type hinting

File: fuel/app/classes/materia/widget/instance.php [hunk 3]
```diff
@@ -180,6 +180,31 @@ public function get_qset($inst_id, $timestamp=false)
...
+		$query = \DB::select()
+			->from('widget_qset')
+			->where('inst_id', $inst_id)
+			->order_by('created_at', 'DESC');
+
+		$results = $query->execute();
+
+		$history = [];
+
+		for ($i = 0; $i < count($results); $i++)
```
Comment (iturgeon) [inline fuel/app/classes/materia/widget/instance.php:194]: prefer foreach

File: fuel/app/config/css.php [hunk 4]
```diff
@@ -62,6 +62,12 @@
 			$static_css.'jquery.dataTables.css',
 			$static_css.'util-question-import.css',
 		],
+		'qset_history' => [
```
Comment (iturgeon) [inline fuel/app/config/css.php:65]: Unless these are large, it may be best to just combine them?

File: fuel/app/tests/api/v1.php [hunk 5]
```diff
@@ -1058,6 +1058,17 @@ public function test_question_set_get()
...
 	}
 
+	public function test_question_history_get()
+	{
+		// ======= AS NO ONE ========
+		try {
+			$output = Api_V1::question_history_get(555);
+			$this->fail("Expected exception HttpNotFoundException not thrown");
+		} catch ( Exception $e) {
+			$this->assertInstanceOf('HttpNotFoundException', $e);
```
Comment (iturgeon) [inline fuel/app/tests/api/v1.php:1068]: Use `$this->expectException(\ HttpNotFoundException::class);` as it's simpler and more semantically specific

File: fuel/app/themes/default/partials/catalog/qset.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+					<th>Saved At</th>
+				</tr>
+			</thead>
+			<tr ng-repeat="save in saves" ng-click="loadSaveData(save.id)">
+				<td>Save #{{saves.length - $index}}</td>
+				<td>{{save.created_at}}</td>
+			</tr>
+		</table>
+	</form>
+	<div class="no-saves" ng-show="saves.length == 0">
```
Comment (iturgeon) [inline fuel/app/themes/default/partials/catalog/qset.php:18]: I think most of the legacy styles are understore based? prefer to match if that's true.

**Review by iturgeon — APPROVED** [review]
🚀

File: fuel/app/classes/controller/api/instance.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	 * Current user must have author/collab access to the widget.
+	 * Requires ?inst_id=<instance id value> to be set
+	 * @return array the qset save history as an array
+	 */
+	public function get_history()
+	{
+		if ( ! $inst_id = Input::get('inst_id')) return $this->response('Requires an inst_id parameter!', 401);
+		if ( ! \Materia\Util_Validator::is_valid_hash($inst_id) ) return $this->response(\Materia\Msg::invalid_input($inst_id), 401);
+		if ( ! ($inst = \Materia\Widget_Instance_Manager::get($inst_id))) return $this->response('Instance not found', 404);
+		if ( ! $inst->playable_by_current_user()) return $this->response(\Materia\Msg::no_login(), 401);
```
Comment (iturgeon) [inline fuel/app/classes/controller/api/instance.php:25]: the name of `playable_by_current_user` seems wrong for this right? should be 'can-edit-instance' (not that that exists)?

https://github.com/ucfopen/Materia/blob/master/fuel/app/classes/controller/widgets.php#L154

Comment (clpetersonucf) [reply fuel/app/classes/controller/api/instance.php]: Resolved and replaced by perm manager's `user_has_any_perm_to` check

**Review by iturgeon — APPROVED** [review]

**Pull request merged**